src/electrum/server.rs

```diff
@@ -1047,3 +1135,159 @@ impl Read for ConnectionStream {
         }
     }
 }
+
+/// This is a slightly modified version of read_until from standard library BufRead trait.
+/// After every read we check if there's a PROXY protocol header at the beginning of the read.
+fn read_until(
+    r: &mut BufReader<ConnectionStream>,
+    delim: u8,
+    buf: &mut Vec<u8>,
+) -> std::io::Result<usize> {
+    let mut read = 0;
+    let mut carry_over_arr = [0_u8; 256];
+    let mut carrying_over = 0;
+    loop {
+        let (done, used) = {
+            let mut available = match r.fill_buf() {
+                Ok(n) => n,
+                Err(ref e) if e.kind() == std::io::ErrorKind::Interrupted => continue,
+                Err(e) => return Err(e),
+            };
+
+            let skipped_count = if carrying_over > 0 {
+                process_carry_over(&mut carrying_over, &mut carry_over_arr, &mut available)
+            } else {
+                // Added to read_until impl: Try parsing PROXY headers after every read.
+                let (_, skipped_count) = parse_proxy_headers(&mut available, 0);
+                skipped_count
+            };
+
+            match memchr::memchr(delim, available) {
+                Some(i) => {
+                    buf.extend_from_slice(&available[..=i]);
+                    (true, i + 1 + skipped_count)
+                }
+                None => {
+                    // Added: carry over
+                    carrying_over = available.len();
+                    carry_over_arr[..carrying_over].copy_from_slice(available);
+
+                    buf.extend_from_slice(available);
+                    (false, available.len() + skipped_count)
+                }
+            }
+        };
+        r.consume(used);
+        read += used;
+        if done || used == 0 {
+            return Ok(read);
+        }
+    }
+}
+
+fn proxy_header_to_source_socket_addr(p_header: ProxyHeader) -> Option<SocketAddr> {
+    match p_header {
+        ProxyHeader::Version1 {
+            addresses: version1::ProxyAddresses::Ipv4 { source, .. },
+        } => Some(SocketAddr::V4(source)),
+        ProxyHeader::Version1 {
+            addresses: version1::ProxyAddresses::Ipv6 { source, .. },
+        } => Some(SocketAddr::V6(source)),
+        ProxyHeader::Version2 {
+            addresses: version2::ProxyAddresses::Ipv4 { source, .. },
+            ..
+        } => Some(SocketAddr::V4(source)),
+        ProxyHeader::Version2 {
+            addresses: version2::ProxyAddresses::Ipv6 { source, .. },
+            ..
+        } => Some(SocketAddr::V6(source)),
+        _ => None,
+    }
+}
+
+fn parse_proxy_headers(
+    buf: &mut &[u8],
+    electrum_proxy_depth: usize,
+) -> (Option<SocketAddr>, usize) {
+    trace!("Starting parse PROXY headers: {:?}", &buf[..12]);
```

Comment:
this trace panics on RPC disconnection (presumably `buf` is empty in that case?)
```
thread 'reader-12' panicked at 'range end index 12 out of range for slice of length 0', src/electrum/server.rs:1212:51
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I can reliably reproduce the panic by running electrs with `-vvvv` and either using the "test connection" feature in Sparrow wallet, or manually connecting and disconnecting the RPC.

Reply:
Resolved

Reply:
Also, I just realized that `panic = 'abort'` was set so all my cool catch_unwind business was ignored....... lol